walkthroughs/week-5-tdd/project/src/main/java/com/google/sps/FindMeetingQuery.java

```diff
@@ -14,10 +14,128 @@
+
+    for (int counter = 0; counter < eventsWithAttendees.size(); counter++) {
+      Event currentEvent = eventsWithAttendees.get(counter);
+
+      // If the array is not empty, then it means there was one event before this one. 
+      if (!availableTimes.isEmpty()) {
+        Event previousEvent = eventsWithAttendees.get(counter - 1);
+
+        // Remove the last added timerange to consider the next event. 
+        availableTimes.remove(availableTimes.size() - 1);
+
+        // If one event fully contains another (nested event), there are 2 possible times: before and after the longest event. 
+        if (previousEvent.getWhen().end() > currentEvent.getWhen().start()
+            && previousEvent.getWhen().end() > currentEvent.getWhen().end()) {
+          availableTimes.add(TimeRange.fromStartEnd(previousEvent.getWhen().end(), TimeRange.END_OF_DAY, true));
+          break;
+        }
+
+        // If events have no conflict, there are 3 possible times: before the first one, between the 2 events and after the second one.
+        long durationBetweenEvents = currentEvent.getWhen().start() - previousEvent.getWhen().end(); 
+        if (request.getDuration() <= durationBetweenEvents) 
+          availableTimes.add(TimeRange.fromStartEnd(previousEvent.getWhen().end(), currentEvent.getWhen().start(), false));
+
+        // If the end of the event is not the end of the day, then the time in between is a possible time. 
+        if (currentEvent.getWhen().end() != TimeRange.END_OF_DAY + 1) {
+          availableTimes.add(TimeRange.fromStartEnd(currentEvent.getWhen().end(), TimeRange.END_OF_DAY, true));
+        }
+      }
+
+      // If there were no previous times, then this event is the first one of the day. The possible times are before and after the event.
+      else {
+        // If the first event of the day starts at the beginning of the day, there are no availabilities before it. 
+        if (currentEvent.getWhen().start() != TimeRange.START_OF_DAY) 
+          availableTimes.add(TimeRange.fromStartEnd(TimeRange.START_OF_DAY, currentEvent.getWhen().start(), false));
+  
+        if (currentEvent.getWhen().end() != TimeRange.END_OF_DAY + 1)
+          availableTimes.add(TimeRange.fromStartEnd(currentEvent.getWhen().end(), TimeRange.END_OF_DAY, true));
+      }
+    }
+    return availableTimes;
+  }
+
+  //Return the events common to both required and optional attendees
+  public Collection<TimeRange> getCommonTimes(List<Event> eventsWithRequiredAttendees,
+      List<Event> eventsWithOptionalAttendees, MeetingRequest request) {
+    
+    Collection<TimeRange> requiredAttendeesAvailabilities = getPossibleTimes(eventsWithRequiredAttendees, request);
+    Collection<TimeRange> optionalAttendeesAvailabilities = getPossibleTimes(eventsWithOptionalAttendees, request);
+    Collection<TimeRange> commonAvailabilities = new ArrayList<>();
+
+    if (optionalAttendeesAvailabilities.isEmpty())
```

Comment:
Not sure if this condition is quite right - what if the required attendees are busy all morning, and optional attendees are busy all afternoon? My impression is that the condition should be: if there are no common availabilities, then just use the required attendee's available times. So it perhaps should be a few lines lower (once you have the common availabilities) and adjusted a bit?